### PR TITLE
feat: evaluate api

### PR DESCRIPTION
**Pull request by Icemap**
Issue: https://github.com/pingcap/autoflow/issues/376

1. Added Create Evaluate Task API and Called Celery as Background;
2. Added a New Queue `evaluation` in the Celery and Changed the Rest Task to Queue `default`;
3. Added Evaluate Task List/Summary/Item Detail APIs.
4. Added Evaluation Dataset List/Create/Delete/Update APIs, a caveat is the CSV file is only can be used in the Create API.
5. Added Evaluation Dataset Item List/Create/Delete/Update APIs.

### CODE REVIEW

File: backend/app/api/admin_routes/evaluation/evaluation_dataset.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+            eval_list = df.to_dict(orient='records')
+            # create evaluation dataset items
+            evaluation_data_list = [EvaluationDatasetItem(
+                query=item["query"],
+                reference=item["reference"],
+                retrieved_contexts=[],  # TODO: implement this after we can retrieve contexts
+                extra={k: item[k] for k in item if k not in must_have_columns},
+            ) for item in eval_list]
+
+    evaluation_task = EvaluationDataset(
```
Comment (Mini256) [inline backend/app/api/admin_routes/evaluation/evaluation_dataset.py:86]: evaluation_dataset?

File: backend/app/api/admin_routes/evaluation/evaluation_dataset.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    session: SessionDep,
+    user: CurrentSuperuserDep
+) -> EvaluationDatasetItem:
+    evaluation_dataset_item = session.get(EvaluationDatasetItem, evaluation_dataset_item_id)
+
+    if not evaluation_dataset_item:
+        raise HTTPException(
+            status_code=status.HTTP_404_NOT_FOUND,
+            detail="Evaluation dataset item not found",
+        )
```
Comment (Mini256) [inline backend/app/api/admin_routes/evaluation/evaluation_dataset.py:208]: Can we wrap this duplicate code in the `evaluation_dataset_repo.must_get`? like: 

https://github.com/pingcap/autoflow/blob/98b8686497f567773bd9f874966f6e7a2aff41f3/backend/app/repositories/embedding_model.py#L47

File: backend/app/api/admin_routes/evaluation/evaluation_dataset.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    evaluation_dataset_item.retrieved_contexts = updated_evaluation_dataset_item.retrieved_contexts
+    evaluation_dataset_item.extra = updated_evaluation_dataset_item.extra
+    evaluation_dataset_item.evaluation_dataset_id = updated_evaluation_dataset_item.evaluation_dataset_id
+
+    session.commit()
+
+    return evaluation_dataset_item
+
+
+@router.get("/admin/evaluation/dataset/{evaluation_dataset_id}/dataset-item")
```
Comment (Mini256) [inline backend/app/api/admin_routes/evaluation/evaluation_dataset.py:221]:
```suggestion
@router.get("/admin/evaluation/datasets/{evaluation_dataset_id}/items")
```

Please follow existing API specifications and use plural noun in the path.

`dataset_items` simpify to `items`.

File: backend/app/api/admin_routes/models.py [hunk 4]
```diff
@@ -1,7 +1,9 @@
+from typing import Optional
 from uuid import UUID
 from pydantic import BaseModel
 
 from app.api.admin_routes.embedding_model.models import EmbeddingModelItem
+from app.models import EvaluationTask
```
Comment (Mini256) [inline backend/app/api/admin_routes/models.py:6]: Are these changes necessary?

File: backend/app/api/admin_routes/evaluation/evaluation_task.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+    session.add(evaluation_task)
+    session.commit()
+
+    add_evaluation_task.delay(evaluation_task.id)
+
+    return evaluation_task
+
+
+@router.get("/admin/evaluation/task-summary/{evaluation_task_id}")
```
Comment (Mini256) [inline backend/app/api/admin_routes/evaluation/evaluation_task.py:81]:
```suggestion
@router.get("/admin/evaluation/tasks/{evaluation_task_id}/summary")
```

File: backend/app/api/admin_routes/evaluation/evaluation_task.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    stmt = (
+        select(EvaluationTask)
+        .where(EvaluationTask.user_id == user.id)
+        .order_by(desc(EvaluationTask.id))
+    )
+    return paginate(session, stmt, params)
+
+
+@router.get("/admin/evaluation/task/all-items/{evaluation_task_id}")
+def list_evaluation_task(
```
Comment (Mini256) [inline backend/app/api/admin_routes/evaluation/evaluation_task.py:160]:
```suggestion
@router.get("/admin/evaluation/tasks/{evaluation_task_id}/items")
def list_evaluation_task_items(
```

File: backend/app/tasks/evaluate.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        SemanticSimilarity(embeddings=evaluator_embeddings)
+    ]
+
+    try:
+        eval_result = evaluate(dataset=ragas_dataset, metrics=metrics, raise_exceptions=True, show_progress=False)
+
+        logger.debug(f"eval_result to_pandas")
+        result_list = eval_result.to_pandas().to_dict(orient='records')
+        logger.debug(f"result list {result_list}")
+        if len(result_list) == 1:
```
Comment (Mini256) [inline backend/app/tasks/evaluate.py:131]:
```suggestion
        if len(result_list) == 0:
           raise Exception(f"Item {evaluation_item.id} cannot get evaluation from ragas")
```

**Review by Mini256 — CHANGES_REQUESTED** [review]

File: backend/app/api/admin_routes/evaluation/evaluation_dataset.py (hunk 1 above)
Comment (Icemap) [reply backend/app/api/admin_routes/evaluation/evaluation_dataset.py:86]: Good catch.

File: backend/app/api/admin_routes/evaluation/evaluation_dataset.py (hunk 2 above)
Comment (Icemap) [reply backend/app/api/admin_routes/evaluation/evaluation_dataset.py:208]: It doesn't have a repository for evaluation tables, use a generic function instead.

File: backend/app/api/admin_routes/evaluation/evaluation_dataset.py (hunk 3 above)
Comment (Icemap) [reply backend/app/api/admin_routes/evaluation/evaluation_dataset.py:221]: Sure, but only for listing APIs.

Comment (Icemap) [reply backend/app/api/admin_routes/evaluation/evaluation_dataset.py:221]: And for `items`'s simplification, I want to adhere to the same name as the other APIs, rather than a short name.

File: backend/app/api/admin_routes/models.py (hunk 4 above)
Comment (Icemap) [reply backend/app/api/admin_routes/models.py:6]: Good catch, optimized.

File: backend/app/api/admin_routes/evaluation/evaluation_task.py (hunk 5 above)
Comment (Icemap) [reply backend/app/api/admin_routes/evaluation/evaluation_task.py:81]: Changed, but for only one task, it will be the `/admin/evaluation/task/{evaluation_task_id}/summary`

File: backend/app/tasks/evaluate.py (hunk 7 above)
Comment (Icemap) [reply backend/app/tasks/evaluate.py:131]: Nope, the logic here is not correct. It only added **one** evaluation item for ragas at once. So the expectation of the length of the return is exactly 1.

File: backend/app/api/admin_routes/evaluation/evaluation_task.py (hunk 6 above)
Comment (Icemap) [reply backend/app/api/admin_routes/evaluation/evaluation_task.py:160]: Ditto

**Comment by Icemap** [comment]
Updated, PTAL again, thanks. @Mini256

File: backend/app/tasks/evaluate.py (hunk 7 above)
Comment (Mini256) [reply backend/app/tasks/evaluate.py:131]: `if len(result_list) != 1:` is ok, the main suggestion is to unify the handling of errors in except statements

File: backend/app/api/admin_routes/evaluation/evaluation_task.py (hunk 5 above)
Comment (Mini256) [reply backend/app/api/admin_routes/evaluation/evaluation_task.py]: Whether to use singular or plural forms is a controversial topic, but in most APIs I’ve encountered, it’s common practice to consistently use the plural form. Additionally, since our existing code already follows this name convention, new code should also follows the same convention.

- https://www.reddit.com/r/AskProgramming/comments/15rxwfd/singular_or_plural_nouns/
- https://stackoverflow.com/questions/6845772/should-i-use-singular-or-plural-name-convention-for-rest-resources

GitHub API exmaple:
- https://docs.github.com/en/rest/checks/runs?apiVersion=2022-11-28#get-a-check-run

The existing API:
https://github.com/pingcap/autoflow/blob/1a369a65a1aa7a11bda22d92d3643533980419fc/backend/app/api/admin_routes/chat_engine.py#L26

File: backend/app/api/admin_routes/evaluation/evaluation_dataset.py (hunk 2 above)
Comment (Mini256) [reply backend/app/api/admin_routes/evaluation/evaluation_dataset.py]: Why not create one?

File: backend/app/models/evaluation_task.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    )
+
+    dataset_id: int = Field(nullable=True)
+
+    evaluation_items: List["EvaluationItem"] = SQLRelationship(back_populates="evaluation_task")
+
+    __tablename__ = "evaluation_tasks"
+
+
+class EvaluationItem(UpdatableBaseModel, table=True):
```
Comment (Mini256) [inline backend/app/models/evaluation_task.py:43]:
```suggestion
class EvaluationTaskItem(UpdatableBaseModel, table=True):
```
Would this be more reasonable?

Comment (Icemap) [reply backend/app/models/evaluation_task.py]: Sure, changed. Affected almost all files, PTAL again. @Mini256

File: backend/app/api/admin_routes/evaluation/evaluation_dataset.py (hunk 2 above)
Comment (Icemap) [reply backend/app/api/admin_routes/evaluation/evaluation_dataset.py]: Pluralitas non est ponenda sine necessitate.

File: backend/app/api/admin_routes/evaluation/evaluation_task.py (hunk 5 above)
Comment (Icemap) [reply backend/app/api/admin_routes/evaluation/evaluation_task.py]: Got it, I'm persuaded.

File: backend/app/models/evaluation_task.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    chat_engine: str = Field(max_length=255)
+    status: EvaluationStatus = Field(sa_column=Column(String(32), nullable=False))
+    query: str = Field(sa_column=Column(Text))
+    reference: str = Field(sa_column=Column(Text))
+    response: str = Field(sa_column=Column(Text))
+    retrieved_contexts: list[str] = Field(default=[], sa_column=Column(JSON))
+    extra: dict = Field(default={}, sa_column=Column(JSON))
+    error_msg: str = Field(sa_column=Column(Text, nullable=True))
+    factual_correctness: Optional[float] = Field(nullable=True)
+    semantic_similarity: Optional[float] = Field(nullable=True)
```
Comment (Mini256) [inline backend/app/models/evaluation_task.py:54]: I have some concerns about the extensibility of this data model:
- How should we handle other evaluation metrics if we have more?
- How to handle user-defined custom metrics?

Although it may be a bit early to consider, table schema migration costs are relatively high.

langfuse's data model may be of some help with us:
https://langfuse.com/docs/scores/data-model

File: backend/app/tasks/evaluate.py (hunk 7 above)
Comment (Icemap) [reply backend/app/tasks/evaluate.py:131]: We want to save the error message. Does this code snippet meet your requirements?
```python
if len(result_list) != 1:
    evaluation_task_item.error_msg = f"Item {evaluation_task_item.id} cannot get evaluation from ragas"
    evaluation_task_item.status = EvaluationStatus.ERROR
else:
    logger.debug(f"result {result_list[0]}")
    evaluation_task_item.factual_correctness = result_list[0][FactualCorrectness.name]
    evaluation_task_item.semantic_similarity = result_list[0][SemanticSimilarity.name]
    evaluation_task_item.status = EvaluationStatus.DONE
```

File: backend/app/models/evaluation_task.py (hunk 9 above)
Comment (Icemap) [reply backend/app/models/evaluation_task.py:54]: It is too early to consider. If we need to add another metric, the effort will be way more than adding a field in the table. And before that, we should fix the retrieved_contexts first. For the user-defined custom metrics, we cannot support this version. PTAL of the `ragas.metrics` package.

File: backend/app/tasks/evaluate.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+        "user_input": evaluation_task_item.query,
+        "reference": evaluation_task_item.reference,
+        "response": evaluation_task_item.response,
+    }]
+    logger.debug(f"Response data {evaluation_task_item.response}")
+
+    ragas_dataset = EvaluationDataset.from_list(ragas_list)
+    logger.debug(f"Dataset {ragas_dataset.to_pandas().head()}")
+
+    evaluator_llm = LangchainLLMWrapper(ChatOpenAI(model="gpt-4o"))
```
[the remark this answers is not in the document]
Comment (Icemap) [reply backend/app/tasks/evaluate.py:115]: It aligns with the command-line version. https://github.com/pingcap/autoflow/blob/main/backend/app/evaluation/evals.py#L26

File: backend/app/tasks/evaluate.py (hunk 7 above)
Comment (Mini256) [reply backend/app/tasks/evaluate.py]: I means:

```python
try:
  if len(result_list) != 1:
    raise Exception(f"Item {evaluation_item.id} cannot get evaluation from ragas")

  logger.debug(f"result {result_list[0]}")
  evaluation_task_item.factual_correctness = result_list[0][FactualCorrectness.name] 
  evaluation_task_item.semantic_similarity = result_list[0][SemanticSimilarity.name]
  evaluation_task_item.status = EvaluationStatus.DONE
  ...
except Exception as e:
  logger.error(f"Failed to evaluate item #{evaluation_item.id}, error: {e}")
  evaluation_item.error_msg = traceback.format_exc()
  evaluation_item.status = EvaluationStatus.ERROR
  ...
```

File: backend/app/tasks/evaluate.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Mini256) [reply backend/app/tasks/evaluate.py:115]: Get!

[the remark this answers is not in the document]
Comment (Icemap) [reply backend/app/tasks/evaluate.py:115]: I tested LlamaIndexLLMWrapper, and it worked. I'll change to the LlamaIndex version. Not only just `LlamaIndexLLMWrapper`, but also `LlamaIndexEmbeddingsWrapper`.

File: backend/app/tasks/evaluate.py (hunk 7 above)
Comment (Icemap) [reply backend/app/tasks/evaluate.py]: Eh! Sorry for my stupid questions. XD
Updated.

File: backend/app/models/evaluation_task.py (hunk 9 above)
Comment (Mini256) [reply backend/app/models/evaluation_task.py:54]: I perfered to store the metrics value in another table like:

```python
class EvaluationTaskItemScore(SQLModel, table=true):
   id: int
   name: varchar(40) # maybe `factual_correctness`, `semantic_similarity`, `faithfulness` and more ...
   value: float
   evaluation_task_item_id: int
   evaluation_task_id: int
```

Just a suggestion and reminder of extensibility, the hard code columns way is also ok for me.

cc:@wd0517 @sykp241095 What do you think?

File: backend/app/models/evaluation_dataset.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+    evaluation_dataset_id: int = Field(foreign_key="evaluation_datasets.id", nullable=True)
+    evaluation_dataset: "EvaluationDataset" = SQLRelationship(
+        back_populates="evaluation_data_list",
+        sa_relationship_kwargs={
+            "lazy": "joined",
+            "primaryjoin": "EvaluationDatasetItem.evaluation_dataset_id == EvaluationDataset.id",
+        },
+    )
+    __tablename__ = "evaluation_datasets_items"
```
Comment (Mini256) [inline backend/app/models/evaluation_dataset.py:47]:
```suggestion
    __tablename__ = "evaluation_dataset_items"
```

Comment (Icemap) [reply backend/app/models/evaluation_dataset.py]: Done

File: backend/app/tasks/evaluate.py [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+            session.commit()
+
+
+@retry(stop=stop_after_attempt(2), wait=wait_fixed(5))
+def generate_answer_by_autoflow(messages: list, chat_engine: str) -> (str, str):
+    response = httpx.post(
+        url=settings.TIDB_AI_CHAT_ENDPOINT,
+        headers={
+            "Content-Type": "application/json",
+            "Authorization": f"Bearer {settings.TIDB_AI_API_KEY}",
```
Comment (wd0517) [inline backend/app/tasks/evaluate.py:152]: `TIDB_AI_API_KEY ` is a SecretStr, should use `settings.TIDB_AI_API_KEY.get_secret_value()`

File: backend/app/models/evaluation_task.py (hunk 9 above)
Comment (wd0517) [reply backend/app/models/evaluation_task.py:54]: Langfuse supports user-defined metrics, so storing them in a separate table is required. However, we currently only support a limited set of hardcoded metrics, making a dedicated column for these acceptable for now.

**Comment by wd0517** [comment]
Please use `rye fmt` to format the code.

**Comment by Icemap** [comment]
Sure, but the `rye fmt` makes a bunch of code files that were updated.

**Comment by Mini256** [comment]
🤣 maybe we need to add a GitHub Action to run `rye fmt --check`

**Comment by Icemap** [comment]
> 🤣 maybe we need to add a GitHub Action to run `rye fmt --check`

Or a git hook maybe.

**Review by Mini256 — APPROVED** [review]
